### PR TITLE
Optimize the implementation of error handling in keep alive.

### PR DESCRIPTION
**Pull request by sighingnow**
Enhance #53.

### CODE REVIEW

File: src/KeepAlive.cpp [hunk 1]
```diff
@@ -152,9 +152,18 @@ void etcd::KeepAlive::refresh()
...
-      this->stubs->call->Refresh();
-      // trigger the next round;
-      this->refresh();
+      if (this->continue_next) {
+        auto resp = this->stubs->call->Refresh();
+        if (!resp.is_ok()) {
+          throw std::runtime_error("Failed to refresh lease: error code: " + std::to_string(resp.error_code()) +
+                                   ", message: " + resp.error_message());
+        }
+        if (resp.value().ttl() == -1) {
```
Comment (pyssling) [inline src/KeepAlive.cpp:161]: it seems this is actually 0, not -1 on expiry. Validated this today.

File: src/KeepAlive.cpp [hunk 2]
```diff
@@ -152,9 +152,18 @@ void etcd::KeepAlive::refresh()
...
+          throw std::runtime_error("Failed to refresh lease due to expiration: the new TTL is -1.");
```
Comment (pyssling) [inline src/KeepAlive.cpp:162]: This  throwing of the same exception for both errors is not good. We cannot distinguish in practice between a failure of the connection - after which we should create a new KeepAlive - and expiry of the lease after which we need a new lease.

**Review by pyssling — COMMENTED** [review]
Some comments - I did an alternate implementation as well - you could look at that one and see what you think. It's important to be able to distinguish between the lease expiry and the connection failure.

File: src/KeepAlive.cpp (hunk 1 above)
Comment (sighingnow) [reply src/KeepAlive.cpp:161]: Thanks. Will fixed.

File: src/KeepAlive.cpp (hunk 2 above)
Comment (sighingnow) [reply src/KeepAlive.cpp:162]: >  It's important to be able to distinguish between the lease expiry and the connection failure.

Agree about distinguishing those two kinds of failure, but I would like to throw different kinds of exception (`std::runtime_error` and `std::invalid_argument` or `std::out_of_range`). I don't want to makes the `KeepAlive` so complicate for users. After adding a another handler, there would be too many ctor combinations and dispatch logic.

**Comment by pyssling** [comment]
out_of_range sounds reasonable!

**Pull request merged**